Add generation test

jyegerlehner opened the pull request:
I hacked in a test to check generation. Right now it just tests WaveNetModel.predict_proba(), not yet WaveNetModel.predict_prob_incremental(). I duplicated a bit of code from generate.py to do it; it's a bit ugly.

The only idea I came up with was to check that the power spectrum of the generated waveform matches what we trained it on. So the training signal looks like this:

![sine_train](https://cloud.githubusercontent.com/assets/2138320/19260829/6ffc7d38-8f52-11e6-8e5b-5b3d74e54930.png)

with peaks at the three frequencies we expect.

A successfully generated waveform looks like this:
![sine_generated](https://cloud.githubusercontent.com/assets/2138320/19260729/aae81836-8f51-11e6-8b28-395bc5a52264.png)

So the test just adds up the power in generated waveform's spectrum, and asserts that at least 90% of it is at those three frequencies. 

I had to train longer to get good results, and generation takes a while. I'm afraid our CI test is getting prohibitively-long. Takes a couple minutes even on the GPU. Afraid to see how long it takes on the CI server.

Is this reasonable approach? Suggestions?

And just for fun, [this](https://soundcloud.com/user-731806733/tf-wavenet-unit-test-sinewave) is what our training signal sounds like.

ibab commented:
This is really clever.
Maybe we can get it to converge faster by increasing the learning rate.

jyegerlehner commented:
I spent a fair bit of time tweaking learning rates, sizes of the dilation stacks, momentum, optimizer & cetera trying to reduce training time. Which is not to say what you suggest isn't possible. But the CI server tests only took about 16 minutes. I don't see that as prohibitive. It's not as bad as I feared.

jyegerlehner commented:
I added testing of fast generation. Cleaned up a smidge. I think something that doesn't duplicate some code from generate.py would require some refactoring of code in generate.py that can be called from test_model.py.

I didn't hear any howls of protest, so I'll remove WIP.

lemonzi commented:
Setting the right frequencies for the sinusoids can get rid of the leak to neighboring bins, and we could then tighten that threshold. I believe the frequencies must be set so that the length of the signal is a multiple of the period.

lemonzi commented:
So, `f_n = n / SAMPLE_DURATION`, and no need to use the sampling rate, just use `1, 2, 3, ...` as `time`. And I'd suggest using `cos` for all of them, it's more standard (the phase in the FFT is then 0 for all sinusoids). We can also change the relative phase in some of them in a more difficult test case.

lemonzi commented:
We could maybe move the synthesis and FFT-based evaluation to a separate file, so that `test_model.py` only tests the network. It's growing a bit long. We could also have a `test_model_base.py`, and then define all the subclasses (biases, different settings, etc) in a separate file.

lemonzi commented:
As mentioned in #141 sometimes the error is small but it's there, so we need to be strict. 

I would also test using the power in dB, rather than in linear units. In that case, rather than squaring it just use `20 * np.log10(np.abs...`

jyegerlehner commented:
> As mentioned in #141 sometimes the error is small but it's there, so we need to be strict. 

I got the impression in 141 that training may have just stopped at 4K iterations, as he said he used defaults. So it wasn't perfect yet. The wave file posted sounded to me like it learned the waveform. I don't agree that 141 identifies any kind of bug that this test should detect. This test was originally designed as a smoke test that we can create a model, train, and learn. You pointed out it we needed to close-the-loop by checking that we generate from the trained model. With this PR, it does that too. 

Since the purpose is to check for bugs, not to demonstrate hyper-fidelity in a trained model, I think we do not want to be "strict". 

As for your other refinements, can those not come along in a future refactoring, when someone thinks they are the best use of his time? They sound like a large investment of time for possibly-arguable stylistic benefit. I'd rather have slightly less-than-optimal-test than no test at all.

lemonzi commented:
Yes, I agree it's better to have this as soon as possible to increase code coverage in the tests. Let's merge it as-is. Can you fix the conflict with the RMSProp?

jyegerlehner commented:
Oops I didn't notice the conflicting file; thanks for the pointer.

jyegerlehner commented:
Oops wait, I just fixed the conflicts. Must be from that last merge.

lemonzi commented:
Still doesn't like it 😮

lemonzi commented:
Maybe you need to pull from this repo onto your fork's master and then rebase, I looked at the network graph and you're a commit behind. That commit was yours, but since it re-writes them maybe it didn't pick it up -- or you don't have it merged on your fork..

jyegerlehner commented:
Thanks for the pointers. This stuff is too complicated for my simple mind. I'll work it out eventually.

ibab commented:
Okay, this looks good :)
